build/docker: do not stack `task_error` decorator

slaperche-scality opened the pull request:
**Component**:

buildchain

**Context**: 

When we have an error in the Dockerfile of one of the builder, the task that build the image swallow the error and return a successful status.
Then build moves forward and fails at a later steps with an unrelated error.

The problems come from our use of the `task_error` decorator.
We stack multiple calls of it to handle different error, but this isn't supported and some errors ends up swallowed.

**Summary**:

Change the implementation of `task_error` so that we can specify a list of errors and their handler in a single call.

**Acceptance criteria**: 

- Delete the builder image: `docker rmi metalk8s-build`
- Delete the on-disk witness: `rm _build/metalk8s-build:latest`
- Insert an error in your Dockerfile, like:

```diff
diff --git a/packages/redhat/Dockerfile b/packages/redhat/Dockerfile
index a48a7aff..c4b13a58 100644
--- a/packages/redhat/Dockerfile
+++ b/packages/redhat/Dockerfile
@@ -11,7 +11,7 @@ RUN sed -i s/@SALT_VERSION@/$SALT_VERSION/ /etc/yum.repos.d/saltstack.repo
 RUN yum install -y \
         createrepo \
         epel-release \
-        rpm-build \
+        rpm-build
         rpmdevtools \
         rpmlint \
         yum-utils \
```

- run `./doit.sh _build_container` and you should see something like:

```
-- MKDIR        _build
.  CHECK CMD    skopeo
.  IMG BUILD    metalk8s-build:latest
TaskError - taskid:_build_container
400 Client Error: Bad Request ("Dockerfile parse error line 15: unknown instruction: RPMDEVTOOLS")

########################################
_build_container <stdout>:
```
---

Closes: #2395

slaperche-scality committed:
build/docker: do not stack `task_error` decorator

We used to stack calls to the `task_error` decorator, one for each error
we wanted to handle.
While this looked nice, it was buggy.

The decorators after the first one were wrapping a decorated function,
not the original one.

Consequently, if we had more than one lever of decorators and the
function raised an error that wasn't handled by the last decorator then
the error was lost:
- the targeted decorator was handling its error as expected and was
  returning a `TaskError`
- but then, the next decorator was seeing a normal return value (a
  TaskError), not an exception and thus `None` was returned => the error
  was lost on the way.

To solve this issue, we've changed the implementation of `task_error`:
we now have a single call that takes a dict mapping an error type to its
handler.

Closes: #2395
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/bugfix/2395-fix-task-error-decorator` with contents from `w/2.3/bugfix/2395-fix-task-error-decorator`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/bugfix/2395-fix-task-error-decorator origin/development/2.4
 $ git merge origin/w/2.3/bugfix/2395-fix-task-error-decorator
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/bugfix/2395-fix-task-error-decorator
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2395-fix-task-error-decorator` into
`development/2.0`
* [w/2.1/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.1` 
* [w/2.2/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.2` 
* [w/2.3/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.3` 
* [w/2.4/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.4` 
* [w/2.5/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.5` 
* [w/2.6/bugfix/2395-fix-task-error-decorator](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2395-fix-task-error-decorator) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

slaperche-scality committed:
build/version: fix image digest

Some digest were outdated and this went unnoticed thanks to the bug we
fix in the previous commit.

Refs: #2395
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
